gitgud/skills/level_builder.py

```diff
@@ -80,11 +81,24 @@ def test_passed(self):
     def test_failed(self):
         default_fail()
 
+    def mark_complete(self):
+        file_operator = operations.get_operator()
+        file_operator.mark_level_complete(self)
+
+    def mark_visited(self):
+        file_operator = operations.get_operator()
+        level_progress = file_operator.get_level_progress(self)
+        if level_progress == "unvisited":
```

Comment:
Even though the net effect is the same, I think this would be better as a not equal to statement, since we don't care if it's unvisited, we care that it's not `complete` or `partial` because we don't want to override either of those

Reply:
So, you'd rather it be expressed as `if not level_progress in {"complete", "partial", "visited"}`? I always understood this function to mean "if the level is not yet visited (`unvisited`), mark the level as visited".

Reply:
As stated elsewhere, `Operator` should handle the completion hierarchy, so feel free to implement that in a single function that lays out all of the hierarchy.

`operator.mark_progress(level, progress)`

```
progress_dict = {"complete":3, "partial":2, "visited":1, "unvisited": 0}
if progress_dict[progress] > progress_dict[old_progress]:
    # mark it
```
This is probably more explicit than either way. Let me know if you think this could be improved on or if there are issues with it

Reply:
Yeah, that's basically what I'm doing right now, except with lists (not yet pushed):
```
    def mark_level(self, level, status):
        progress_data = self.read_progress_file()
        heirarchy = [
            "unvisited", "visited", "partial", "complete"
        ]
        current_progress = self.get_level_progress(level)
        if heirarchy.index(status) > heirarchy.index(current_progress):
            progress_data[level.skill.name].update(
                {level.name: status}
            )
            self.update_progress_file(progress_data)
```

Reply:
Yes, that's fine